### PR TITLE
Refactor to isolate BlueZ code

### PR DESCRIPTION
**Pull request by beroset**
Refactor to hide more of the details of the bluezclient code within the
code and narrow the interface to only what is externally needed.

### CODE REVIEW

File: asteroidsyncserviced/bluez/bluezclient.h [hunk 1]
```diff
@@ -57,11 +56,7 @@ private slots:
 
 private:
     bool isAsteroidOSWatch(const QStringList uuids) const;
-    QDBusConnection m_dbus;
-    DBusObjectManagerInterface m_bluezManager;
-    BluezAdapter1 *m_bluezAdapter = nullptr;
-    FreeDesktopProperties *m_bluezAdapterProperties = nullptr;
-
+    BluezClientImpl *bt = nullptr;
```
Comment (FlorentRevest) [inline asteroidsyncserviced/bluez/bluezclient.h]: If you make this a std::unique_ptr it will be deleted automatically and you don't need a BluezClient destructor

File: asteroidsyncserviced/platforms/ubuntutouch/ubuntuplatform.h [hunk 2]
```diff
@@ -2,6 +2,7 @@
 #define UBUNTUPLATFORM_H
 
 #include <QDBusContext>
+#include <QDBusInterface>
```
Comment (FlorentRevest) [inline asteroidsyncserviced/platforms/ubuntutouch/ubuntuplatform.h:5]: Why is this here ? same for the new include in ubuntuplatform.cpp

File: asteroidsyncserviced/main.cpp [hunk 3]
```diff
@@ -22,10 +22,12 @@
...
 
-#ifdef UBUNTU_TOUCH_PLATFORM
-#include "platforms/ubuntutouch/ubuntuplatform.h"
+#if UBUNTU_TOUCH_PLATFORM
+#include "ubuntuplatform.h"
+using Platform = UbuntuPlatform;
 #elif SAILFISHOS_PLATFORM
-#include "platforms/sailfishos/sailfishplatform.h"
+#include "sailfishplatform.h"
+using Platform = SailfishPlatform;
```
Comment (FlorentRevest) [inline asteroidsyncserviced/main.cpp]: Ok.
But notice that this does not fit in the scope of what your commit describes. Best practices in version control systems is to have atomic commits https://tekin.co.uk/2021/01/how-atomic-commits-make-you-a-better-coder

File: asteroidsyncservice_config.h.in [hunk 4]
```diff
@@ -21,7 +21,7 @@
 #define CONFIG_H
 
 #define VERSION "${asteroidsyncservice_VERSION_MAJOR}.${asteroidsyncservice_VERSION_MINOR}"
-#cmakedefine UBUNTU_TOUCH_PLATFORM
-#cmakedefine SAILFISHOS_PLATFORM
+#cmakedefine01 UBUNTU_TOUCH_PLATFORM
+#cmakedefine01 SAILFISHOS_PLATFORM
```
Comment (FlorentRevest) [inline asteroidsyncservice_config.h.in]: Why is that necesssary ?

File: asteroidsyncserviced/bluez/bluezclient.cpp [hunk 5]
```diff
@@ -19,28 +19,50 @@
...
+#include <stdexcept>
+
+class BluezClientImpl {
+public:
+    BluezClientImpl();
+    QDBusConnection m_dbus;
+    DBusObjectManagerInterface m_bluezManager;
+private:
+    BluezAdapter1 *m_bluezAdapter = nullptr;
+    FreeDesktopProperties *m_bluezAdapterProperties = nullptr;
```
Comment (FlorentRevest) [inline asteroidsyncserviced/bluez/bluezclient.cpp]: It looks like these two private members are never used, you can remove them then.

File: asteroidsyncserviced/bluez/bluezclient.cpp [hunk 6]
```diff
@@ -19,28 +19,50 @@
...
+#include "freedesktop_properties.h"
+#include "bluez_adapter1.h"
+#include "bluez_agentmanager1.h"
 
 #include <QDBusConnection>
 #include <QDBusReply>
 #include <QDebug>
+#include <stdexcept>
+
+class BluezClientImpl {
```
Comment (FlorentRevest) [inline asteroidsyncserviced/bluez/bluezclient.cpp]: The design-pattern you're using here is the "opaque pointer" https://en.wikipedia.org/wiki/Opaque_pointer#C++
The "Impl" suffix does not seem appropriate because this does not actually implement anything. This is just a glorified storage space.
If you like that pattern, what about following the Qt Project's best practices https://wiki.qt.io/D-Pointer ? Typically, in the Qt codebase, this would be called BluezClientPrivate.

File: asteroidsyncserviced/bluez/bluezclient.h (hunk 1 above)
Comment (FlorentRevest) [inline asteroidsyncserviced/bluez/bluezclient.h]: I can't make sense of "bt" ? If you follow Qt practices this would be something like d_ptr.

**Review by FlorentRevest — CHANGES_REQUESTED** [review]

File: asteroidsyncservice_config.h.in (hunk 4 above)
Comment (beroset) [reply asteroidsyncservice_config.h.in]: In the places that these were tested in the code they used #ifdef ... #elif, but that logic was faulty because the first clause tests for definition and the second for value.  This change defines both with either a 1 or 0, making it easier to use with #if .. #elif.

File: asteroidsyncserviced/bluez/bluezclient.cpp (hunk 5 above)
Comment (beroset) [reply asteroidsyncserviced/bluez/bluezclient.cpp]: You are right.  In fact, that also means that we no longer need several files in the bluez directory.

File: asteroidsyncserviced/bluez/bluezclient.cpp (hunk 6 above)
Comment (beroset) [reply asteroidsyncserviced/bluez/bluezclient.cpp]: Thanks, I did not know about that Qt best practice and have now adopted it.

File: asteroidsyncserviced/bluez/bluezclient.h (hunk 1 above)
Comment (beroset) [reply asteroidsyncserviced/bluez/bluezclient.h]: You're right.  Done.

File: asteroidsyncserviced/main.cpp (hunk 3 above)
Comment (beroset) [reply asteroidsyncserviced/main.cpp]: True, but the code won't compile without that change so I closed a previous PR to combine these.  Would you have preferred that I keep the other one (which was *only* this change) and make this PR dependent on that other one?

File: asteroidsyncserviced/platforms/ubuntutouch/ubuntuplatform.h (hunk 2 above)
Comment (beroset) [reply asteroidsyncserviced/platforms/ubuntutouch/ubuntuplatform.h:5]: The interface uses QDBusInterface, but there is no corresponding #include.  Apparently we assume <QDBusContext> or some other header also includes it.  I'll omit it, but I'm not sure of the rationale.

File: asteroidsyncserviced/bluez/bluezclient.h (hunk 1 above)
Comment (beroset) [reply asteroidsyncserviced/bluez/bluezclient.h]: Good point!  Done.

**Comment by beroset** [comment]
Thanks once again for the detailed review and patient comments.

File: asteroidsyncserviced/main.cpp (hunk 3 above)
Comment (FlorentRevest) [reply asteroidsyncserviced/main.cpp]: Notice that I'm talking about commits, not PRs. Commits are units of code changes, PRs are units of code reviews. A PR is a bundle of commits which get reviewed at the same time. PRs disappear in time but commits stay in the commit log, which people read to understand a project.

Also I see you added commits on top of the older one such as "Address review comments and further simplify.". These are virtually impossible to understand when going through the git log of a repository if a PR like this is merged as is.

You can learn how to rewrite a git commit log here https://about.gitlab.com/blog/2020/11/23/keep-git-history-clean-with-interactive-rebase/ and use that to follow the best practices recommended in my first link.

I know this sounds like I'm just trying to be a massive pain in the back but believe me, this is really important. Every day, I stumble upon code and I want to know why it was introduced, I use git blame on a line which tells me when the line was introduced and points to the commit that explains why, along with the rest of the context. If changes are spread across various "fix things" commits and mixed with random other changes, it gets very hard to understand why something was introduced. Also when I find that such a commit was actually bad, it's trivial to use a `git revert` whereas in this situation one needs to untangle all the changes.

---

To save you from having to think about all the things that are going on here, here is an example of how I would, personally, arrange commits in this PR. It doesn't have to be exactly like this, this is just an example.

```
Standardize how UBUNTU_TOUCH_PLATFORM and SAILFISHOS_PLATFORM are defined and checked

These macros were defined without a value, but used with an elif. This sets them to 1 and tests them by value rather than by presence.
```

This would only contain the changes in `asteroidsyncservice_config.h.in` and replace the ifdef with an if here.

Notice that, had this been an individual commit, I would not have had to ask why these changes were introduced.

```
Centralize the definitions of Platform types in one place

This keeps the macro mess more self-contained.
```

This would only contain the changes toward the end of this file and the introduction of "using Platform = " at the top. Notice that the title is easy to read in a git log and the commit contains a justification of why this is beneficial. Justifying your own changes is a great way to make sure you know what you're doing and why. For example, in the first stage of this PR you introduced includes for no reason, if you were to write a commit description for this you would notice that this shouldn't be committed.

```
Move the definitions of PLATFORM_CODE_NAME and PLATFORM_CODE_NAME to the top level CMakeLists.txt

This removes some duplicated code
```

This would only contain the changes to CMakeLists.txt

```
Shorten platform headers include paths

PLATFORM_CODE_DIR is already in the include directories list so we can use a shorter include.
```

This would only contain the change where you remove `platforms/ubuntutouch/` and `platforms/sailfishos/` from the include directives here.

```
Shorten libasteroid headers include paths

../libasteroid is already in the include directories list so we can use a shorter include
```

This would only contain a one-line change in `asteroidsyncserviced/bluez/bluezclient.cpp`

```
Remove dead-code

The `BluezAdapter1` and `FreeDesktopProperties` are no longer used.
```

This would only contain the removal of various files under bluez/ and their listing from `asteroidsyncserviced/CMakeLists.txt`

```
Remove an unused macro

BLUEZ_ADAPTER_IFACE isn't used anywhere
```

This would only contain the change to asteroidsyncserviced/bluez/dbus-shared.h

```
Uniformize coding-style

Two inconsistencies slipped through the BluezClient declaration
```

This would only contain the `slotInterfacesAdded` and `slotInterfacesRemoved` line changes

```
Use the D-Pointer pattern in BluezClient

This isolates private member definitions in a separate class to keep implementation details hidden from source files including bluezclient.h
```

---

If you compare to the current commit log:

```
Refactor to isolate BlueZ code
Refactor to hide more of the details of the bluezclient code within the
code and narrow the interface to only what is externally needed.
```

It should be pretty obvious now that this commit actually does *a lot more* than this. If you put yourself in the shoes of the reviewer or the person who will later track down why a change was introduced, you should understand that a giant commit under a title that does not justify why half of the code is introduced forces them to untangle every line of code themselves and figure out what any of this has to do with hiding details of the bluezclient code.

```
Address review comments and further simplify.
```

Here again, review context will be lost. Also the person who will have spent time understanding the previous commit will now need to forget everything that changed.

```
Add the private interface header.
```

As a side note this means that at the previous commit, code does not build. This effectively breaks bisectability and causes issues in many projects. (which run CI on every commit for example) Luckily for you we are not strict about bisectability in AsteroidOS but it's also a best practice to not break things to fix them in the next commit.

---

Anyway, sorry this got a bit long. I hope it doesn't come out as aggressive, I'm genuinely trying to help and truly appreciate all your work. I only spend time explaining all of this because I believe this will help you, not only in AsteroidOS but in any other project you'll want to contribute to later.

File: asteroidsyncserviced/bluez/bluezclient.h [hunk 7]
```diff
@@ -20,15 +20,12 @@
 #ifndef BLUEZCLIENT_H
 #define BLUEZCLIENT_H
 
+#include "bluez_helper.h"
+#include "bluezclient_p.h"
```
Comment (FlorentRevest) [inline asteroidsyncserviced/bluez/bluezclient.h]: If you include the bluezclient_p header here this kills all the benfits of the d-pointer pattern. Source files including bluezclient.h would transitively also including bluezclient_p.h and therefore pull in all the other includes.

If you want to take advantage of the d-pointer pattern you should use a forward declaration of the BluezClientPrivate type.

File: asteroidsyncserviced/bluez/bluezclient.cpp [hunk 8]
```diff
@@ -18,29 +18,38 @@
...
+BluezClient::BluezClient(QObject *parent):
+    QObject(parent),
+    d_ptr{std::make_unique<BluezClientPrivate>()}
 {
     qDBusRegisterMetaType<InterfaceList>();
     qDBusRegisterMetaType<ManagedObjectList>();
 
-    if (m_bluezManager.isValid()) {
-        connect(&m_bluezManager, SIGNAL(InterfacesAdded(QDBusObjectPath,InterfaceList)),
+    if (d_ptr && d_ptr->m_bluezManager.isValid()) {
```
Comment (FlorentRevest) [inline asteroidsyncserviced/bluez/bluezclient.cpp]: I know you may have learned that dynamic allocation results should always be checked but this is actually an anti pattern in C++. The `!d_ptr` condition will actually never be true because standard C++ allocations throw a `std::bad_alloc` exception if something fails to allocate. Btw no need to catch that exception, if we're ever in a situation where the d_ptr allocation fails the whole thing is already crashing anyway.

File: asteroidsyncserviced/bluez/bluezclient.cpp [hunk 9]
```diff
@@ -100,13 +109,17 @@ BluezClient::BluezClient(QObject *parent):
                 }
             }
         }
+    } 
+    else 
+    {
+        throw std::runtime_error("unable to connect to D-Bus");
```
Comment (FlorentRevest) [inline asteroidsyncserviced/bluez/bluezclient.cpp]: The above `std::bad_alloc` discussion is actually a very good introduction to why I believe you should just completely avoid using exceptions. https://pianomanfrazier.com/post/exceptions-considered-harmful/

Notice that the Qt framework does not use any exception ever, precisely because it's a horrible language construct.

Also here you're throwing an exception but never catching it so instead of gracefully continuing the execution this will cause a hard crash. I recommend you just log something to the standard error output instead.

File: asteroidsyncserviced/bluez/bluezclient.cpp [hunk 10]
```diff
@@ -18,29 +18,38 @@
...
+    if (d_ptr && d_ptr->m_bluezManager.isValid()) {
+        connect(&d_ptr->m_bluezManager, SIGNAL(InterfacesAdded(QDBusObjectPath,InterfaceList)),
                 this, SLOT(slotInterfacesAdded(QDBusObjectPath,InterfaceList)));
 
-        connect(&m_bluezManager, SIGNAL(InterfacesRemoved(QDBusObjectPath,QStringList)),
+        connect(&d_ptr->m_bluezManager, SIGNAL(InterfacesRemoved(QDBusObjectPath,QStringList)),
                 this, SLOT(slotInterfacesRemoved(QDBusObjectPath,QStringList)));
 
-        auto objectList = m_bluezManager.GetManagedObjects().argumentAt<0>();
+        auto objectList = d_ptr->m_bluezManager.GetManagedObjects().argumentAt<0>();
```
Comment (FlorentRevest) [inline asteroidsyncserviced/bluez/bluezclient.cpp]: Could you keep things consistent by using `d_func` here too ? (just like in `BluezClient::pairedWatches`)

File: asteroidsyncserviced/bluez/bluezclient.h [hunk 11]
```diff
@@ -20,15 +20,12 @@
 #ifndef BLUEZCLIENT_H
 #define BLUEZCLIENT_H
 
+#include "bluez_helper.h"
```
Comment (FlorentRevest) [inline asteroidsyncserviced/bluez/bluezclient.h:23]: I believe this is only required in bluezclient.cpp. If not, only include it in the header, not both the header and the compilation unit.

**Review by FlorentRevest — CHANGES_REQUESTED** [review]
Thanks! Keep on, it's nearly there :)

File: asteroidsyncserviced/bluez/bluezclient.h (hunk 11 above)
Comment (beroset) [reply asteroidsyncserviced/bluez/bluezclient.h:23]: It's needed in the header because it contains some typedefs that are needed for the interface.

File: asteroidsyncserviced/bluez/bluezclient.cpp (hunk 10 above)
Comment (beroset) [reply asteroidsyncserviced/bluez/bluezclient.cpp]: Yes, all raw `d_ptr` uses that were convertible to use `d_func()` now do so.

**Comment by FlorentRevest** [comment]
🚀

Thank you! :)

File: asteroidsyncserviced/bluez/bluezclient.cpp (hunk 9 above)
Comment (beroset) [reply asteroidsyncserviced/bluez/bluezclient.cpp]: This is now changed to use qCritical to emit a message instead.
